# Lösungsmodelle entwerfen

![](../images/serenity/serenity8.jpg)

Flow-Design verfolgt einen verhaltensorientierten Ansatz. Lösungen werden von der SOLL-Funktionalität her angegangen. Für Flow-Design lautet die Frage zuerst: Was soll Software tun?

Nicht Datenstrukturen stehen also im Vordergrund, sondern Verhaltensstrukturen. Verhalten konsumiert und produziert Daten. Um zu wissen, wie Daten strukturiert sein sollen, ist also zuerst die Verhaltensproduktion zu klären.

Was wie womit getan werden soll, entsteht dabei zunächst in den Köpfen der Entwickler und auf Papier. Die Codierung ist zurückgestellt, bis für ein klar umrissenes Problem die Lösung als Modell vorliegt.

Ein Modell versteht Flow-Design dabei ganz allgemein so:

> Modell = (Funktion, Beziehung, Funktion)\*

Ein Modell ist eine Menge von Tupeln, in denen je zwei Funktion in Beziehung zueinander gesetzt werden.

Wie auch immer Modelle dargestellt werden, aus ihnen müssen programmiersprachliche Funktionen abgeleitet werden können. In Modellen findet sich keine Logik; sie sind deklarativ. Aber ihre Bausteine stehen für Funktionen, die später in der Codierung mit Logik gefüllt werden.

Beziehungen zwischen Funktionen können unter anderem sein:

* Abhängigkeit: Eine Funktion ruft eine andere auf.
* Sequenz: Eine Funktion steht in einem Fluss vor oder nach einer anderen.
* Parallelität: Eine Funktion kann/muss parallel zu einer anderen ausgeführt werden.
* Aggregation: Zwei Funktionen sind in einem Modul zusammengefasst.
* Gemeinsame Daten: Zwei Funktionen nutzen dieselben Daten.
* Konkretisierung: Zwei Funktionen stehen in einer Vererbungsbeziehung.

Die Deklarativität von Modellen im Flow-Design macht aus, dass ihre Bausteine für angenommene Problemlösungen stehen. Entweder wird dann das Problem an anderer Stelle im Modell durch eine Verfeinerung gelöst. Oder eine im Modell offen gebliebene Annahme wird spätestens in der Codierung mittels Logik erfüllt.

![](images/df1.png)

Bei all dieser Allgemeinheit in der Definition von Modellen hat Flow-Design allerdings einen klaren Favoriten als Modellierungsansatz. Flow-Design steht für die Überzeugung, dass Softwarelösungen zunächst und ganz fundamental Produktionsprozesse sind, die auf Daten arbeiten und also als Datenflüsse modelliert werden können. Der Entwurf im Flow-Design beginnt daher gemeinhin mit einer Datenflussmodellierung.

## Mit Datenflüssen modellieren

Die Datenflussmodellierung beginnt gewöhnlich mit einer in der Analyse ermittelten Interaktion repräsentiert durch ein Nachrichtenpaar und ggf. Seiteneffekte. Ausgangspunkt ist eine Funktionseinheit, die Verhalten zeigen soll in Form einer klar definierten Datentransformation. Sie stellt die Wurzel des Lösungsansatzes dar. Sie repräsentiert die angenommen erfüllten Anforderungen.

![](images/df2.png)

Wie aufwändig die Lösung eines so gestellten Verhaltensproblems sein mag, ist seiner formalen Darstellung nicht anzusehen. Das ist aber auch nicht nötig, wie sich zeigen wird. Wichtiger ist die einfachheit der Darstellung und ihre Klarheit im Sinne einer Codierung.

### 0-dimensionale Datenflüsse

Eine Funktionseinheit, die Input in Output überführt, ist das kleinstmögliche Datenfluss-Modell. Sie entspricht einem Punkt im dreidimensionalen Raum, ist also 0-dimensional.

![](images/df3.png)

#### Notation

So leichtgewichtig die Notation für solch eine Repräsentanz einer Problemlösung sein mag, so sind doch ein paar Hinweise angebracht:

* Eine Funktionseinheit dargestellt als "Blase" oder Kreis oder Ellipse, also als "runde Form" steht für etwas, das getan werden soll; sie löst ein Problem durch ihre Transformation von Input zu Output. Deshalb ist ihre Benennung mit einem Verb oder einer Verbphrase sinnvoll. Ein Imperativ muss es nicht sein, aber darf es natürlich. Zudem sollte die Benennung relevant für das Umfeld sein, also orientiert an der Domänensprache statt technisch. Außerdem - und vielleicht sogar am wichtigsten - sollte die Bennung keinen Hinweis auf die Lösung(simplementation) enthalten. Die wird im Modell nur gewünscht, nicht gewusst. **Es geht im Modell ums Was und Warum, nicht ums Wie.**
* Datenflüsse werden durch Pfeile zur und von der Funktionseinheit angezeigt. Pfeile werden im Flow-Design ausschließlich für diesen Zweck benutzt! Pfeile stellen also keine Abhängigkeiten dar, sondern unidirektionale Datenkanäle, deren Quellen und Senken einander nicht kennen.
* Was fließt, wird immer (!) an Pfeilen in Klammern kurz beschrieben. Auch hier geht es vor allem ums Was, nicht ums Wie. Daten werden mit Substantiven benannt. Meist geschieht das in Kleinschreibung und ohne Datentyp, weil der offensichtlich ist. Aber es kann auch ein Datentyp explizit hinzugefügt werden. Oder Daten werden in Großschreibung benannt und stehen damit selbst für einen Datentyp.

![](images/df4.png)

##### Zustandsbehaftete Funktionseinheiten

Funktionseinheiten arbeiten vor allem auf dem Input, der in sie einfließt. Eintreffender Input triggert ihr Verhalten. Aber Funktionseinheiten sind nicht per se *pur*. Flow-Design hat kein Ideal von "pure functional units" wie die Funktionale Programmierung "pure functions" favorisiert. Flow-Design empfiehlt Zustandslosigkeit bzw. Seiteneffektfreiheit, wo sie sinnvoll und machbar sind. Doch wenn sich das (zunächst) als unintuitiv erweisen sollte, sind Zustand und Seiteneffekte völlig akzeptable - sollten jedoch angezeigt werden.

![](images/df5.png)

Wie Zustand genau realisiert wird, ist wieder nicht Sache des Modells. Es soll vor allem klar gemacht werden, dass es eine Einflussgröße für die Transformation gibt, die über Nachrichten hinweg relevant ist.

Wenn Zustand an gewisser Stelle vermieden werden soll, kann er auch in den Fluss extrahiert werden:

![](images/df6.png)

Dasselbe gilt für Seiteneffekte. Die werden über einen API hergestellt, der auch in spezifischeren Funktionseinheiten up-/downstream enger gekapselt werden könnte:

![](images/df7.png)

##### Datenströme

Die meisten Funktionseinheiten transformieren eine Input-Nachricht in eine Output-Nachricht. Dabei ist es unerheblich, ob Input oder Output aus einem Datum bestehen oder ein Feld sind, eine Liste oder eine andere Sammlung von vielen Daten.

![](images/df8.png)

Wenn Sammlungen (*collections*) fließen, kann deren Typ explizit angeben werden - z.B. `(dateiname[])`, `(dateinamen:List<string>)`. Im Allgemeinen ist das jedoch schon zu viel Implementationsdetail für ein Modell. Deshalb ziegt Flow-Design ein schlichtere Angabe für "mehrere Daten" vor, die "in einem Schwung" ein-/ausfließen: `(dateiname*)`. Der Datenangabe ist einfach nur unmittelbar und in der Klammer ein Sternchen nachzustellen.

![](images/df9.png)

Eine oft zu vernachlässigende, am Ende jedoch sehr interessante Eigenschaft von Datenflüssen im Gegensatz zu Kontrollflüssen ist jedoch, dass alle Funktionseinheiten im Grunde *gleichzeitig* aktiv sind. Nur weil Daten produziert sind und ausfließen, heißt das nicht, dass eine Funktionseinheit die Kontrolle abgibt. Sie kann auch weiterarbeiten, während der Output zu einer anderen Funktionseinheit fließt und gleichzeitig dort verarbeitet wird. Produzenten und Konsumenten sind grundsätzlich nebenläufig und asynchron - und nur, wenn das nicht wichtig ist, werden sie sequenziell und synchron implementiert.

Die grundsätzliche Nebenläufigkeit hat nun zur Folge, dass eine Funktionseinheit nicht nur einmal für einen Input einen Output produzieren kann, sondern mehrfach. Dann entsteht keine *collection* von Nachrichten, sondern ein *stream*. Jedes Element in solch einem  Strom ist eine eigene Nachricht. Angezeigt wird das im Modell durch ein Sternchen *hinter* der Klammer, z.B. `(dateiname)*`.

Ströme sind nur relevant zu markieren als Output. Die Anlieferung von Input erfolgt im Grunde immer als Strom. Notiert wird das nur nicht speziell; es ist normal. Auf jede Nachricht im Input-Strom reagiert eine Funktionseinheit dann.

![](images/df10.png)

#### Übersetzung

"Bubbles don't crash" - das ist wahr und deshalb ist es Flow-Design wichtig, dass die "Bubbles" seiner Modelle möglichst einfach in Code übersetzt werden können, der crashen kann, um Modell-Ideen zu überprüfen.

Die naheliegende Übersetzung einer Funktionseinheit ist die in eine Funktion:

![](images/df11.png)

In [ ]:
int CalcAverage(IEnumerable<int> werte) {
    return werte.Sum() / werte.Count();
}

display($"Durchschnitt von [1,5,9,3]: {CalcAverage(new[]{1,5,9,3})}")

Das funktioniert gut, solange für jeden Input-Wert ein Output-Wert bzw. eine Output-Collection erzeugt wird. Und auch mit Tupeln funktioniert es, wenn die Programmiersprache Tupel direkt unterstützt oder man gewillt ist, für ein Tupel einen Datentypen zu definieren:

In [ ]:
(T head, IEnumerable<T> tail) Split<T>(IEnumerable<T> list)
    => (list.First(), list.Skip(1));

var result = Split(new[]{2,7,9,3});
display($"head: {result.head}");
foreach(var e in result.tail) display($"  tail element: {e}");

Funktionen sind natürliche "Objekte" im Sinne von Alan Kay, weil sie für eine empfangene Nachricht - *dass* man sie aufruft und mit welchen Parametern man sie aufruft - ein Resulat herstellen, ohne zu wissen, woher sie aufgerufen wurden und wie das Resultat benutzt wird. Das PoMO wird eingehalten.

##### Datenströme

Allerdings können Funktionen für jedes eintreffende Nachricht, d.h. für jeden Aufruf, nur einmal ein Resultat herstellen (auch wenn das womöglich aus mehreren Elementen besteht). Was aber, wenn Input in einen Strom von Resultaten umzuwandeln ist?

Hier bricht die Übersetzung der "Blase" in eine Funktion. Bei genauerem Hinsehen ist es stattdessen so, dass die Punkte, wo Pfeile auf eine Funktionseinheit treffen und wo sie sie verlassen, getrennt übersetzt werden:

* Ein eintreffender Pfeil wird in eine Funktion mit Parametern übersetzt.
* Ein ausgehender Pfeil wird in ein Funktionsresultat oder eine *continuation* übersetzt.

Ein Funktionsresultat ist dann möglich, wenn nur einmal ein Output generiert wird. Andernfalls muss Output über einen Funktionszeiger an die Umwelt weitergeschoben werden. Dieser Funktionszeiger muss eine *Prozedur* beschreiben (Funktion ohne Rückgabewert), deren Name nicht auf weitere downstream Verarbeitung des Output hindeutet, da sonst das PoMO verletzt würde.

![](images/df12.png)

In [42]:
IEnumerable<string> SplitIntoWords(string text)
    => text.Split(new[]{' ', '\n', '\t'}, StringSplitOptions.RemoveEmptyEntries);

var words = SplitIntoWords("the quick brown fox");
foreach(var w in words) display(w);

the

quick

brown

fox

In [43]:
void EnumerateWords(string text, Action<string> onWord)
    => text.Split(new[]{' ', '\n', '\t'}, StringSplitOptions.RemoveEmptyEntries).ToList()
           .ForEach(onWord);

EnumerateWords("humpty dumpty sat on a wall",
    word => display(word));

humpty

dumpty

sat

on

a

wall

Die Codierung eines Output-Datenkanals mit einer continuation ist die universellere. Aber sie ist weniger intuitiv und umständlicher. Deshalb zieht Flow-Design die Übersetzung mit Funktionsresultaten vor, wo sie offensichtlich ist oder auch machbar durch Sprachkonstrukte wie Iteratoren (z.B. `IEnumerable<>` in C#).

Streams sind Mengen von einzelnen Werten. Wo sie naheliegend und nützlich im Modell sind, sollten sie angemessen in der Übersetzung repräsentiert werden. Das ist mit continuations, also Funktionszeigern in vielen Programmiersprachen möglich, wenn auch ein wenig gewöhnungsbedürftig. Aber die Gewöhnung lohnt sich: die Übersetzung von Modellen wird einfacher, die Testbarkeit des Codes steigt.

##### Zustand

Zustand bzw. Ressourcenzugriff sind Abhängigkeiten einer Funktionseinheit. Sie werden als Abhängigkeiten übersetzt. Zustand ist dabei eine globale Variable im aggregierenden Modul einer Funktion:

![](images/df13.png)

In [44]:
class Accumulator {
    private int _value;
    
    public int Add(int a) {
        _value += a;
        return _value;
    }
}

var accu = new Accumulator();
display($"add 1: {accu.Add(1)}");
display($"add 2: {accu.Add(2)}");
display($"add 3: {accu.Add(3)}");

add 1: 1

add 2: 3

add 3: 6

### 1-dimensionale Datenflüsse

Einzelne Funktionsheiten werden in Modellen zu Datenflüssen mehrere Funktionseinheiten zusammengesetzt. Flow-Design nennt das auch "zusammenstecken" oder "verdrahten" in Anlehnung an elektronische Schaltungen, die übrigens ganz natürlich dem PoMO folgen. Aus 0-dimensionalen Modellen werden so 1-dimensionale: viele punktuelle Funktionseinheiten zusammen ergeben einen linearen Fluss.

![](images/df14.png)

Wenn der Output einer Funktionseinheit in der Form dem Input einer anderen entspricht, können beide in eine producer-consumer Beziehung gebracht werden. Sie spannen dann einen Transformationsfluss auf.

Die Übersetzung eines 1-dimensionalen Flusses ist trivial: die Funktionen der einzelnen Funktionseinheiten...

In [45]:
// Funktionseinheiten
int[] Map(string roman)
    => roman.ToCharArray()
            .Select(romanDigit => romanDigit switch {
                        'I' => 1, 'V' => 5, 'X' => 10,
                        'L' => 50, 'C' => 100, 'D' => 500,
                        'M' => 1000
                    })
            .ToArray();

int[] Negate(int[] values) {
    var negatedValues = (int[])values.Clone();
    for(var i=0; i<negatedValues.Length-1; i++)
    if (negatedValues[i]<negatedValues[i+1])
        negatedValues[i] *= -1;
    return negatedValues;
}

int Sum(IEnumerable<int> values) {
    var sum = 0;
    foreach(var v in values) sum += v;
    return sum;
}

 ...werden schlicht nacheinander aufgerufen. Der Compiler gibt Feedback, ob die Bausteine eines Datenflusses zueinander passen.

In [46]:
// Datenfluss
var values = Map("XIV");
values = Negate(values);
var decimalNumber = Sum(values);

display($"XIV={decimalNumber}");

XIV=14

In dieser Übersetzung verliert das Modell des Datenflusses zwar seine Eigenschaft der grundsätzlichen Asynchronizität seiner Transformationsschritte, doch das ist in den meisten Fällen unerheblich. Datenflüsse sind auch Datenflüsse, wenn sie synchron operieren. Das Modell hat trotzdem einen hohen Wert durch seine Abstraktion.

Die Funktionseinheiten, die im 1-dimensionalen Datenfluss verdrahtet werden, sind (zunächst) Operationen. Der Datenfluss stellt ihre Integration dar: alle einzelnen Transformationen bilden in spezifischer Weise zusammengesteckt eine größere Transformation.

#### Datenströme

Solche Übersetzung von 1-dimensionalen Flüssen als Lösungen von größeren Problemen in Form einer Abfolge von Lösungen kleinerer Probleme funktionieren auch für Ströme von Einzeldaten statt Collections.

Datenströme sind nützlich, wenn...

* ...die Zahl der zu erzeugenden Output-Datenelemente unbekannt und wahrscheinlich groß ist. In dem Fall soll vielleicht vermieden werden, alle Datenelemente in einer Collection im Hauptspeicher zu sammeln.
* ...die Generierung aller Output-Datenelemente erhebliche Zeit in Anspruch nimmt. In dem Fall sollen schon generierte Output-Daten schon downstream in Verarbeitung gehen, während weitere beschafft werden.
* ...ein Datenfluss besser zu verstehen und/oder zu übersetzen ist als Strom von Einzeldaten.

Als Beispiel dafür mag die Traversierung eines Verzeichnisbaumes dienen, in dem Dateien eines bestimmten Typs gesucht werden, um sie zu analysieren. Zu ermitteln ist die Gesamtanzahl der Worte in relevanten Dateien.

![](images/df15.png)

Die Zahl der zu prüfenden Dateien ist in so einem Szenario womöglich schwer abzuschätzen und/oder die Lösung ist leichter "zu denken", wenn die Analyse sich auf einzelne Dateien konzentriert.

Zuerst die Übersetzung der Funktionseinheiten, der Bausteine des Flusses. Jede ist nur eine kleine Operation mit wenigen Zeilen Code:

In [47]:
void EnumerateFiles(string path, Action<string> onFilename) {
    // Die Logik hier ist aufwändiger als nötig aus didaktischen Gründen, um einen Output-Strom zu erzeugen.
    var filenames = System.IO.Directory.GetFiles(path);
    foreach(var f in filenames) {
        onFilename(f);
    }
    
    var subdirectories = System.IO.Directory.GetDirectories(path);
    foreach(var d in subdirectories)
        EnumerateFiles(d,
              onFilename);
}

void FilterByEndOfFilename(string filename, string pattern, Action<string> onFilename) {
    if (filename.EndsWith(pattern))
        onFilename(filename);
}

string Load(string filename) 
    => System.IO.File.ReadAllText(filename);

string[] SplitIntoWords(string text)
    => text.Split(new[]{' ', '\n', '\t'}, StringSplitOptions.RemoveEmptyEntries);

void Print(string filename, int numberOfWords) {
    display($"{filename} with {numberOfWords} words");
}

Und dann die Übersetzung des Flusses. Die mag ein wenig gewöhnungsbedürftig sein in der Schachtelung der Lambda-Funktionen als continuations. Doch das verliert sich, wenn man es ein paar Mal geschrieben und gelesen hat. Der Gewinn der Gewöhnung liegt dann darin, eine geradlinige Übersetzung für Datenflüsse mit Strömen zu haben, die in jedem Schritt gut testbar ist.

In [49]:
EnumerateFiles("..",
    filename => FilterByEndOfFilename(filename, ".ipynb",
          filename => {
              var text = Load(filename);
              var words = SplitIntoWords(text);
              Print(filename, words.Length);
          })
);

../anforderungen.ipynb with 4475 words

../prozess.ipynb with 1827 words

../anforderung-logik-lücke.ipynb with 3276 words

../index.ipynb with 125 words

../entwurf/entwurf_1.ipynb with 3485 words

../entwurf/.ipynb_checkpoints/entwurf_1-checkpoint.ipynb with 3388 words

../codierung/arbeitsorganisation.ipynb with 37 words

../codierung/.ipynb_checkpoints/arbeitsorganisation-checkpoint.ipynb with 37 words

../analyse/slicing.ipynb with 49 words

../analyse/.ipynb_checkpoints/slicing-checkpoint.ipynb with 49 words

../anatomie/radikale_oo.ipynb with 3736 words

../anatomie/ioda.ipynb with 3216 words

../anatomie/dimensionen.ipynb with 2475 words

../anatomie/.ipynb_checkpoints/radikale_oo-checkpoint.ipynb with 3736 words

../anatomie/.ipynb_checkpoints/dimensionen-checkpoint.ipynb with 2475 words

../anatomie/.ipynb_checkpoints/ioda-checkpoint.ipynb with 3216 words

../.ipynb_checkpoints/anatomie_1-checkpoint.ipynb with 3734 words

../.ipynb_checkpoints/anforderung-logik-lücke-checkpoint.ipynb with 3276 words

../.ipynb_checkpoints/anatomie_3-checkpoint.ipynb with 3372 words

../.ipynb_checkpoints/prozess-checkpoint.ipynb with 1827 words

../.ipynb_checkpoints/index-checkpoint.ipynb with 115 words

../.ipynb_checkpoints/anatomie_2-checkpoint.ipynb with 2549 words

../.ipynb_checkpoints/anforderungen-checkpoint.ipynb with 4473 words

`EnumerateFiles` erzeugt für eine Input-Nachricht einen Strom von Output-Nachrichten. `FilterByEndOfFilename` hingegen hat die Aufgabe, Input zu verschlucken, falls er nicht dem Filterkriterium entspricht. Das klingt sehr unterschiedlich, ist letztlich jedoch dasselbe: Der Strom dient in beiden Fällen zur Ausgabe einer unbekannten Zahl von Elementen, das können 1 oder 1000 sein - oder auch gar keines.

Eine Funktion, die ein Resultat mittels `return` liefert, ist gezwungen, eines zu produzieren, selbst wenn das `null` sein sollte. Eine Funktion, die ihr Resultat stattdessen über eine continuation "hinausschiebt", kann auch entscheiden, gar kein Resultat zu liefern! Deshalb ist die Übersetzung von Datenfluss-Funktionseinheiten in Funktionen mit continuation die universellere, wenn auch etwas aufwändigere.

### 2-dimensionale Datenflüsse

Funktionen sind in 0-dimensionalen und 1-dimensionalen Datenflüssen die Container für Logik. Sie komponieren aus "Logikbausteinen" ein Verhalten, für das sie stehen.

Obwohl 1-dimensionale Datenflüsse leicht zu verstehen sind auch in der Übersetzung in Funktionsaufrufsequenzen, skalieren sie nicht. Mehr als 5 oder vielleicht 10 Funktionseinheiten "in einer Reihe" oder Funktionsaufrufe nacheinander, sind aus mehreren Gründen nicht praktikabel:

* Im Modell wie im Code sind so lange Sequenzen trotz ihrer formalen Einfachheit nicht mehr gut zu überblicken.
* Die Gefahr, dass Datenflüsse das SLA verletzen, steigt mit jeder weiteren Funktionseinheit.
* Die Wahrscheinlichkeit, dass es in Datenflüssen Gruppen von Funktionseinheiten mit höherer Kohäsion untereinander gibt als zu anderen, steigt mit jeder weiteren Funktionseinheit.

Schon zum oben stehenden Datenfluss zur Analyse von Dateien kann gefragt werden, ob er für das Verständnis eine optimale Länge hat.

#### Integrationen als Bedeutungseinheiten

Einerseits hat ein Leser bei dem 1-dimensionalen Datenfluss "alles auf einmal im Blick". Alle Operationen stehen in einer Reihe. Andererseits muss ein Leser sich die Bedeutung des Ganzen aus den Bedeutungen der Teile erst zusammenreimen. Das ist bei 5 Funktionseinheiten schwieriger als bei 4, 3 oder 1.

Was tut der Datenfluss? Kann das mit 1 Funktionseinheit ausgedrückt werden? Selbstverständlich!

![](images/df16.png)

Die eine Funktionseinheit integriert nun explizit die bisherigen Operationen.

Oder gibt es im Datenfluss noch Teile, die wiederum unter einem Begriff zusammengefasst werden können, um ihnen eine eigene Bedeutung zu geben? Beispiele dafür wären:

* Alles außer der Ausgabe könnte als Kernlogik zusammengefasst werden. Die Ausgabe eines Ergebnisses ist etwas ganz anderes, als das Ergebnis herzustellen. Dadurch würde die Ergebnisherstellung separat von der Benutzerschnittstelle testbar.
* Die Beschaffung der relevanten Dateinamen könnte zusammengefasst werden. Die Traversierung des Dateisystems ist etwas ganz anderes, als mit einer Datei umzugehen.
* Laden und Zerlegen einer Datei könnte zusammengefasst werden. Eine Datei als Text zu beschaffen, statt als Byteblock, ist schon eine Abstraktion, die der .NET Framework zur Verfügung stellt. Warum nicht den Dateizugriff weiter abstrahieren und eine Datei als Collection von Worten repräsentieren?

Wie man sich entscheidet bei den Zusammenfassungen, ist weniger wichtig, als dass Zusammenfassungen zur Verbesserung der Verständlichkeit und der Testbarkeit möglich und einfach sind. Die eine beste Zusammenfassung wird es wohl nicht geben. Da Integrationen von Teilflüssen jedoch billig herzustellen und auch wieder aufzulösen sind, kann die Integrationshierarchie jederzeit umgestellt werden.

Verständnis verändert sich mit der Zeit: Es wird aufgebaut durch Beschäftigung mit Code - aber es wird auch wieder abgebaut, wenn die Beschäftigung abnimmt. Eine Struktur, die dem heutigen Verständnis angemessen ist, passt womöglich nicht mehr optimal zum morgigen Verständnis.

In dieser Hinsicht unterliegen Modelle und Code trotz ihrer Immaterialität einem Verschleiß. Ohne, dass Code "sich bewegte" oder verändert würde, verliert er an "Verständlichkeitsqualität". Das hat zwei Konsequenzen:

* Die Diskussion um die allerbeste Struktur sollte begrenzt werden. Die allerbeste Struktur, selbst wenn sie für heute gefunden werden könnte, ist morgen nicht mehr die allerbeste. Nach dem Herstellungszeitpunkt verliert jede auf Verständlichkeit getrimmte Struktur sofort und automatisch an Wert. Warum also bis aufs Letzte optimieren?
* Strukturen für Verständlichkeit sollten gewartet werden. D.h. sie sollten proaktiv in angemessenen Wartungsinverallen aufgesucht und dem aktuellen Verständnisstand angepasst werden. Sonst besteht die Gefahr, dass sie zur Unzeit als stark erodiert erfahren werden und viel Mühe beim Verständnis machen.

Die obigen Fragen zu möglichen Integrationen sind daher nur Beispiele dafür, dass und wie möglichen Verständlichkeitsverbesserungen nachgespürt werden kann. Nach angemessenem Diskurs muss eine Entscheidung getroffen werden. Die könnte z.B. so aussehen:

![](images/df17.png)

Ein Datenflussmodell dehnt sich damit sichtbar in 2 Dimensionen aus:

* In der ersten Dimension wird Verhalten im Datenfluss erzeugt: am Anfang einfließerder Input wird schrittweise in am Ende ausfließenden Output transformiert.
* In der zweiten Dimension findet Abstraktion durch Komposition statt. Kompositionen - Operation wie Integration - fassen verschiedene Teilverhalten zu einem größeren, neuen Gesamtverhalten zusammen. Komposition ist die Abstraktion, die Bequemlichkeit in der Nutzung herstellt.

Kompositionen ändern am Verhalten natürlich nichts. Das Verhalten wird immer nur durch Logik (und Verteilung) beeinflusst. Aber Kompositionen ändern an der Verständlichkeit etwas. Deshalb ist auch der Einwand, dass in einer Integrationshierarchie Operationen und Integrationen womöglich nur einmal verwendet werden, unerheblich. Wenn eine Komposition der Verständlichkeit oder Testbarkeit dient, dann muss sie nicht gleichzeitig der kurzfristigen Produktivität dienen. Das ist nämlich der Zweck von Wiederverwendung.

#### Übersetzung

Sind Integrationen für einen Fluss gefunden, lassen sie sich leicht übersetzen: wie Operationen werden sie zu Funktionen. Das ist einerseits sehr nützlich und intuitiv. Andererseits hat es aber auch einen Nachteil.

Integrationen in Funktionen zu übersetzen, liegt nahe, weil die Darstellung im Modell sich nicht von Operationen unterscheidet. Auch weil während des Modellierungsprozesses nicht klar ist, ob Blätter in einer Integrationshierarchie Operationen bleiben oder weiter verfeinert werden, d.h. "in Integrationen umklappen", wäre eine unterschiedliche Darstellung bzw. Codierung hinderlich.

Andererseits verführt die Übersetzung von Integrationen in Funktionen dazu, Logik in sie "hineinzuschmuggeln". Das würde dem IOSP widersprechen und mindestens die Testbarkeit senken.

Letztlich überwiegt jedoch für Flow-Design die Einfachheit. Integrationen werden also zu Funktionen:

In [72]:
void PrintWordsInFiles(string path, string pattern) {
    CountWordsInFiles(path, pattern,
        Print);
}

void CountWordsInFiles(string path, string pattern, Action<string,int> onFile) {
    EnumerateRelevantFiles(path, pattern,
        filename => {
            var words = LoadWordsFromFile(filename);
            onFile(filename, words.Length);
        });
}

void EnumerateRelevantFiles(string path, string pattern, Action<string> onFilename) {
    EnumerateFiles("..",
        filename => FilterByEndOfFilename(filename, pattern,
                      onFilename));
}

string[] LoadWordsFromFile(string filename) {
    var text = Load(filename);
    return SplitIntoWords(text);
}


PrintWordsInFiles("..", ".ipynb");

../anforderungen.ipynb with 4475 words

../prozess.ipynb with 1827 words

../anforderung-logik-lücke.ipynb with 3276 words

../index.ipynb with 125 words

../entwurf/entwurf_1.ipynb with 4778 words

../entwurf/.ipynb_checkpoints/entwurf_1-checkpoint.ipynb with 4776 words

../codierung/arbeitsorganisation.ipynb with 37 words

../codierung/.ipynb_checkpoints/arbeitsorganisation-checkpoint.ipynb with 37 words

../analyse/slicing.ipynb with 49 words

../analyse/.ipynb_checkpoints/slicing-checkpoint.ipynb with 49 words

../anatomie/radikale_oo.ipynb with 3736 words

../anatomie/ioda.ipynb with 3216 words

../anatomie/dimensionen.ipynb with 2475 words

../anatomie/.ipynb_checkpoints/radikale_oo-checkpoint.ipynb with 3736 words

../anatomie/.ipynb_checkpoints/dimensionen-checkpoint.ipynb with 2475 words

../anatomie/.ipynb_checkpoints/ioda-checkpoint.ipynb with 3216 words

../.ipynb_checkpoints/anatomie_1-checkpoint.ipynb with 3734 words

../.ipynb_checkpoints/anforderung-logik-lücke-checkpoint.ipynb with 3276 words

../.ipynb_checkpoints/anatomie_3-checkpoint.ipynb with 3372 words

../.ipynb_checkpoints/prozess-checkpoint.ipynb with 1827 words

../.ipynb_checkpoints/index-checkpoint.ipynb with 115 words

../.ipynb_checkpoints/anatomie_2-checkpoint.ipynb with 2549 words

../.ipynb_checkpoints/anforderungen-checkpoint.ipynb with 4473 words

#### Jo-Jo Modellierung

Im Flow-Design beginnt die Modellierung der Lösung eines Problem außen und beim Gesamtverhalten. Von dort setzt sich durch Zerlegung von größeren in kleinere Probleme nach innen fort. Sie ist verhaltensorientiert und schrittweise verfeinernd, d.h. rekursiv absteigend: Am Anfang steht eine Funktionseinheit als vorläufige Operation. Deren Transformation wird dann jedoch zerlegt in Schritte, die durch nächste vorläufige Operationen in einem Fluss vollständig geleistet wird, so dass die vorherige Funktionseinheit zu einer Integration wird. Und immer so weiter für jedes Blatt im 2-dimensionalen Datenflussbaum - bis man am Ende nicht mehr weiß, wie eine vorläufige Operation weiter zerlegt werden soll. Dann wird aus der vorläufigen Operation eine endgültige.

Zumindest verläuft die Modellierung des Verhaltens idealerweise so. In der Praxis jedoch zeigt sich, dass nicht immer outside-in, top-down eine Lösung erkannt wird. Manchmal steht schon am Anfang ein spätere Operation fest und die wird schrittweise an die Wurzel-Funktionseinheit bottom-up angebunden. Oder nach einem ersten Durchstich von der Wurzel bis zu den Operationen scheint es verständlichkeitsfördernd, weitere Integrationsebenen einzuziehen. Oder Kind-Knoten im Kompositionsbaum, die unter verschiedenen Eltern-Knoten hängen, werden horizontal verschoben.

![](images/df18.png)

Flow-Design ist insofern weder strickt top-down, noch bottom-up, sondern geht in beide Richtungen vor oder schwingt auch mal seitwärts. Das passende Bild für ihre Bewegungsrichtung ist mithin ein Jo-Jo: der geht immer wieder runter und hoch oder bricht auch mal aus und kommt am Ende zur Ruhe. 

#### Logik in der Integration

Das IOSP ist glasklar: Eine Funktionseinheit ist entweder Operation und steht für Verhaltenserzeugung durch Logik. Oder sie ist eine Integration und stellt nur einen Fluss zwischen anderen Funktionseinheiten her, wofür keine Logik nötig ist.

Damit formuliert das Prinzip ein Soll, ein Ideal. Und dieses Ideal kann technisch auch immer implementiert werden.

Allerdings ist es eben ein Ideal. Das bedeutet, in seiner Reinheit kann es anderen Zielen auch mal im Wege stehen.

Flow-Design sieht dieses und andere Prinzipien daher pragmatisch. Ja, es soll stets das Ideal im Sinne von Korrektheit und Evolvierbarkeit angestrebt werden. Dabei soll aber auch Augenmaß bewahrt werden. IOSP, PoMO usw. für langfristig hohe Produktivität sind auszubalancieren mit Maßnahmen für funktionale und nicht-funktionale Qualitäten.

Für Flow-Design sind deshalb gelegentliche "Spuren von Logik" in eigentlich integrierenden Funktionseinheiten akzeptabel. Nicht wünschenswert, aber ok - wenn man denn weiß, was man da tut.

Meistens handelt es sich bei diesen Logik-Einsprengseln um Kontrollstrukturen oder triviale selbsterklärende API-Aufrufe.

Als Beispiel mag nochmal die obige Dateianalyse dienen. Mit der Erlaubnis zu etwas Logik in der Integration könnte die Übersetzung anders aussehen:

In [81]:
// Schmutzige Integration
void PrintWordsInFiles_v2(string path, string pattern) {
    foreach(var r in CountWordsInFiles_v2(path, pattern)) // Kontrollstruktur😱
        Print(r.filename, r.numberOfWords);
}

// Schmutzige Integration
IEnumerable<(string filename, int numberOfWords)> CountWordsInFiles_v2(string path, string pattern) {
    foreach(var f in EnumerateRelevantFiles_v2(path, pattern)) { // Kontrollstruktur😱
        var words = LoadWordsFromFile_v2(f);
        yield return (f, words.Length);
    };
}

// Schmutzige Integration
IEnumerable<string> EnumerateRelevantFiles_v2(string path, string pattern) {
    foreach(var f in EnumerateFiles_v2(path)) // Kontrollstruktur😱
        if (IsRelevantFile(f, pattern)) // Kontrollstruktur😱
            yield return f;
}

// Schmutzige Integration
string[] LoadWordsFromFile_v2(string filename) {
    var text = System.IO.File.ReadAllText(filename); // API-Aufruf😱
    return SplitIntoWords(text);
}


// Geänderte Operation für den Aufruf in einer schmutzigen Integration
IEnumerable<string> EnumerateFiles_v2(string path) {
    // Die Logik hier ist aufwändiger als nötig aus didaktischen Gründen, um einen Output-Strom zu erzeugen.
    var filenames = System.IO.Directory.GetFiles(path);
    foreach(var f in filenames) {
        yield return f; // Statt über eine Continuation wird ein Strom jetzt mittel Iterator erzeugt🤔
    }
    
    var subdirectories = System.IO.Directory.GetDirectories(path);
    foreach(var d in subdirectories)
        foreach(var f in EnumerateFiles_v2(d))
            yield return f; //🤔
}

// Geänderte Operation für den Aufruf in einer schmutzigen Integration
bool IsRelevantFile(string filename, string pattern)
    => filename.EndsWith(pattern);


PrintWordsInFiles_v2("..", ".ipynb");

../anforderungen.ipynb with 4475 words

../prozess.ipynb with 1827 words

../anforderung-logik-lücke.ipynb with 3276 words

../index.ipynb with 124 words

../entwurf/entwurf_1.ipynb with 5776 words

../entwurf/.ipynb_checkpoints/entwurf_1-checkpoint.ipynb with 5776 words

../codierung/arbeitsorganisation.ipynb with 37 words

../codierung/.ipynb_checkpoints/arbeitsorganisation-checkpoint.ipynb with 37 words

../analyse/slicing.ipynb with 49 words

../analyse/.ipynb_checkpoints/slicing-checkpoint.ipynb with 49 words

../anatomie/radikale_oo.ipynb with 3736 words

../anatomie/ioda.ipynb with 3216 words

../anatomie/dimensionen.ipynb with 2475 words

../anatomie/.ipynb_checkpoints/radikale_oo-checkpoint.ipynb with 3736 words

../anatomie/.ipynb_checkpoints/dimensionen-checkpoint.ipynb with 2475 words

../anatomie/.ipynb_checkpoints/ioda-checkpoint.ipynb with 3216 words

../.ipynb_checkpoints/anatomie_1-checkpoint.ipynb with 3734 words

../.ipynb_checkpoints/anforderung-logik-lücke-checkpoint.ipynb with 3276 words

../.ipynb_checkpoints/anatomie_3-checkpoint.ipynb with 3372 words

../.ipynb_checkpoints/prozess-checkpoint.ipynb with 1827 words

../.ipynb_checkpoints/index-checkpoint.ipynb with 124 words

../.ipynb_checkpoints/anatomie_2-checkpoint.ipynb with 2549 words

../.ipynb_checkpoints/anforderungen-checkpoint.ipynb with 4473 words

In jeder einzelnen Methode hält sich die Unsauberkeit in Bezug auf das IOSP in Grenzen. Die Methoden bleiben auch klein. Solange beides gegeben ist, steht aus Sicht von Flow-Design ein wenig "Schmutz" nichts entgegen. Wichtiger als die absolute Befolgung des Prinzips sind die Qualitäten, die es befördert, hier: **Verständlichkeit und Testbarkeit. Solange diese Qualitäten auch trotz (oder wegen?) gewisser Unsauberkeit vorhanden sind, ist etwas Logik in einer Integrationsfunktion ok.**

[Solche "Schmutzreste"](https://ccd-school.de/2017/02/kontrollstrukturen-in-der-integration/) sind vor allem simple Schleifen, auch mal eine Fallunterscheidung oder ein trivialer API-Aufruf. **Wichtig ist, dass diese Logik so einfach ist, dass sie selbst nicht getestet werden muss.** Das ist z.B. hier der Fall:

```csharp
if (IsRelevantFile(f, pattern))
    yield return f;
```

Die `if`-Anweisung selbst kann nicht falsch geschrieben werden; der Compiler würde das beanstanden. Die eigentliche Logik steckt in der Funktion `IsRelevantFile`, zu der nun eine funktionale Abhängigkeit besteht. Diese Logik ist durch die Auslagerung jedoch leicht testbar.

Wenn ein `if` in solcher Weise zur Lesbarkeit einer Integration beiträgt... Warum dann darauf verzichten? Flow-Design will mit seinen Prinzipien leiten, aber nicht darauf reiten.

Ähnlich sieht es für `foreach` aus, mit dem der Code auf Continuations verzichten kann, weil C# mit `yield return` es erlaubt, sehr einfach Iteratoren nach und nach mit Elementen zu füllen.

#### Offen für Erweiterung, geschlossen für Veränderung

Die Vorteile der Einhaltung von IOSP und PoMO liegen zunächst sichtbar in größerer Verständlichkeit und besserer Testbarkeit von Code. Logik wird ganz natürlich in kleine Einheiten "gezwungen". Flow-Design verzichtet ganz bewusst auf die Möglichkeit funktionaler Abhängigkeiten, um Produktivitätsqualitäten verlässlicher herzustellen. [Functional dependencies considered harmful!](https://ralfw.de/2019/07/functional-dependencies-considered-harmful/) Freiwillige Selbstbeschränkung beim Schreiben von Code fördert das spätere und häufigere Lesen und Verändern von Code.

Der positive Effekt des IOSP entstammt zunächst einer Zuspitzung des SRP: Integration und Operation sind formale Verantwortlichkeiten. Während üblicherweise das SRP inhaltlich bzw. auf das Verhalten angewandt wird, konzentriert sich das IOSP auf die Struktur von Code. Deshalb kann die Einhaltung des IOSP auch leicht überprüft werden, selbst wenn ein Leser mit einer Programmiersprache oder dem angestrebten Verhalten oder der Domäne nur wenig vertraut ist.

Weiterhin befördert das IOSP die Einhaltung des SLA. Wenn Funktionen keine funktionalene Abhängigkeiten enthalten, dann liegt das, was sie zusammenfassen (Komposition) im Abstraktionsgrad näher beieinander.

Und schließlich kann man das PoMO als eine Steigerung des *Interface Segregation Principle (ISP)* auffassen. Denn wenn downstream Konsumenten upstream Produzenten über einen Funktionszeiger (continuation) bekannt gemacht werden, dann entspricht das einer Abhängigkeit von einem Interface mit nur einer Methode. Es findet eine Funktionsinjektion in eine Funktion im Bedarfsfall statt.

Damit aber nicht genug! IOSP und PoMO sind auch ganz im Sinne des *(Polymorphic) Open/Closed Principle (OCP)*! [Robert C. Martin definiert das](https://drive.google.com/file/d/0BwhCYaYDn8EgN2M5MTkwM2EtNWFkZC00ZTI3LWFjZTUtNTFhZGZiYmUzODc1/view) wie folgt:

> Modules that conform to the open-closed principle have two primary attributes.
> 1. They are “Open For Extension”. This means that the behavior of the module can be extended. That we can make the module behave in new and different ways as the requirements of the application change, or to meet the needs of new applications.
> 2. They are “Closed for Modification”. The source code of such a module is inviolate. No one is allowed to make source code changes to it.

Flow-Design übersetzt "module" dabei allerdings in "Funktionseinheit" bzw. "Funktion" als eigentlich Verhalten erzeugende Strukturbestandteile. Und Flow-Design konkretisiert, was denn nicht verändert werden sollte: Logik. Logik ist schwer zu verstehen und schwer zu testen. Daher sollte man die Finger davon lassen, sobald sie einmal tut, was sie soll. Veränderungen für neues Verhalten sollten nicht durch Veränderung (modification) von Logik hergestellt werden - sondern durch neue, separate Logik (extension).

Dem leistet Flow-Design Vorschub durch den Verzicht auf funktionale Abhängigkeiten. Operationen sieht Flow-Design als geschlossen an; ihre Logik sollte soweit möglich nicht angetastet werden. Integrationen hingegen sind für Flow-Design offen. In Integrationen können Erweiterungen des Verhaltens leicht "zwischengeschaltet" werden.

Zunächst ein Szenario ohne IOSP: die Umwandlung römischer Zahlen in dezimale noch ohne Berücksichtigung der Subtraktionsregel.

In [85]:
int FromRoman(string romanNumber) {
    var decimalNumber = 0;
    for(var i=0; i<romanNumber.Length; i++) {
        switch(romanNumber[i]) {
            case 'I': decimalNumber += 1; break;
            case 'V': decimalNumber += 5; break;
            case 'X': decimalNumber += 10; break;
            // ...
        }
    }
    return decimalNumber;
}


display($"XVI={FromRoman("XVI")}")

XVI=16

Was tun, um auch die Anforderungen der Subtraktionsregel zu erfüllen? Die Logik von `FromRoman` muss *verändert* werden! Das ist riskant/schwierig.

Anders liegt der Fall, wenn eine Lösung - auch wenn sie klein ist - sauber modelliert und mit IOSP im Kopf übersetzt wird. Sie sähe dann z.B. so aus:

In [91]:
int FromRoman_OCP(string romanNumber) {
    var digitValues = Map(romanNumber);
    return digitValues.Sum();
}

int[] Map(string romanNumber)
    => romanNumber.ToCharArray().Select(Map).ToArray();
int Map(char romanDigit) => romanDigit switch {
    'I' => 1,
    'V' => 5,
    'X' => 10
    // ...
};


display($"XVI={FromRoman_OCP("XVI")}");
display($"XIV={FromRoman_OCP("XIV")}");

XVI=16

XIV=16

Noch leistet die Lösung nicht, was sie soll. Doch ohne Veränderung einer Operation - `Map`, `Sum` - kann das geändert werden mit einer Erweiterung der Integration. Deren Extension besteht im "Dazwischenschieben" eines weiteren Funktionsaufrufs:

In [95]:
int FromRoman_OCP_extended(string romanNumber) {
    var digitValues = Map(romanNumber);
    digitValues = Negate(digitValues); // Erweiterung🤩
    return digitValues.Sum();
}

// Neue Logik, statt veränderter🥳
int[] Negate(IEnumerable<int> digitValues) {
    var negated = digitValues.ToArray();
    for(var i=0; i<negated.Length-1; i++)
        if (negated[i]<negated[i+1])
            negated[i] *= -1;
    return negated;
}


display($"XVI={FromRoman_OCP_extended("XVI")}");
display($"XIV={FromRoman_OCP_extended("XIV")}");

XVI=16

XIV=14

Durch die Trennung von Integration und Operation reduziert Flow-Design den Aufwand, um dem OCP zu dienen. Erweiterbarkeit ist ein grundlegendes Merkmal von Integrationen. Dass man dafür Code im Fluss der Verhaltensherstellung anfassen muss, ist unerheblich. Das Risiko für Schaden (Regression) dadurch ist vergleichsweise klein. Wahrhaft risikoreiche Veränderungen an Logik sind auf sehr überschaubare Operationen begrenzt.

Die erste Frage beim Modellieren eines Lösungsansatzes für neues Verhalten lautet im Flow-Design: Kann das neue Verhalten hergestellt werden, in den lediglich Funktionseinheiten in einem Datenfluss zwischen andere gesetzt werden? Öfter als man denkt, kann darauf mit Ja geantwortet werden. (Und wenn nicht, kann das als Anlass genommen werden, den bisherigen Lösungsansatz nocheinmal zu überdenken, um ihn mittels IOSP näher ans OCP zu bringen.)

#### Stratified Design

Dem Verständnis eines Systems ist zuträglich, wenn man es aus unterschiedlicher Entfernung betrachten kann. Ist man näher dran, kann man Details untersuchen. Ist man weiter weg, gewinnt man Überblick über den Zusammenhang von Teilen.

Geografische Karten erlauben uns, in solcher Weise mit einem Terrain umzugehen. Wir ziehen sie z.B. zur Planung einer Reise in unterschiedlichem Maßstab heran. Eine Karte mit großem Maßstab, z.B. eine Europa-Karte, gibt uns Überblick für eine grobe Routenplanung. Eine topografische Karte hingegen zeigt uns eine eng begrenzte Umgebung mit viel mehr Einzelheiten. Werkzeuge wie Google Maps machen heute den Wechsel des Maßstabs besonders einfach.

Für das Studium von Code ist es ebenfalls nützlich, den Maßstab seiner Darstellung wechseln zu können. Das Ganze sollte grob und im Überblick betrachtet werden können; genauso sollte es möglich sein, schrittweise hinein zu zoomen, um mehr und mehr Details zu sehen.

Große Maßstäbe stehen für ein hohes Abstraktionsniveau, kleine Maßstäbe für ein niedriges. Um hinein und heraus zoomen zu können, muss Code also unterschiedliche Abstraktionsniveaus aufweisen.

Tut er das z.B., wenn er nach dem weit verbreiteten Schichtenmodell strukturiert ist?

Als Beispiel ein kleines Programm, dass die Worte in einer Datei zählt. Es besteht aus drei typischen Schichten:

* Presentationsschicht/Benutzerschnittstelle: Beschafft den Namen der zu verarbeitenden Datei aus der Umgebung und gibt das Ergebnis aus, das es von der Geschäftslogik bekommt.
* Geschäftslogikschicht: Ermittelt das Ergebnis, nachdem es von der Datenzugriffschicht den Dateiinhalt bekommen hat.
* Datenzugriffsschicht: Lädt den Inhalt einer Datei.

/// schichtenmodell für das beispiel (erstmal nur klassennamen) (100)

Funktionale Abhängigkeiten zeigen hier grundsätzlich über die Schichten hinweg von oben nach unten. Ob das kaschiert wird mit DIP/IoC, ist unwesentlich. Und auch Schalen statt Schichten, wie in der Clean Architecture, machen keinen Unterschied. So oder so sind die Aspekte funktional voneinander abhängig zur Laufzeit.

In [102]:
class Presentation {
    public static void CountWords(string filename) {
        var numberOfWords = Business.CountWords(filename);
        Console.WriteLine($"Number of significant words in '{filename}': {numberOfWords}");
    }
}

class Business {
    public static int CountWords(string filename) {
        var words = Dataaccess.Load(filename);
        var significantWords = words.Where(w => w.Length > 2);
        return significantWords.Count();
    }
}

class Dataaccess {
    public static IEnumerable<string> Load(string filename) {
        var text = System.IO.File.ReadAllText(filename);
        return text.Split(new[]{' ', '\t', '\n', '\r'}, StringSplitOptions.RemoveEmptyEntries);
    }
}


Presentation.CountWords("samples/poem.txt");

Number of significant words in 'samples/poem.txt': 72


Der Code ist das Terrain. Das Schichtenmodell ist eine Karte für dieses Terrain. Sie abstrahiert von den Details der Logik und sogar von den einzelnen Methoden. Das ist durchaus nützlich. Und das ganze Programm mit einem Betriebssystemprozess als Host abstrahiert sogar noch von den Klassen als Feinheiten der Implementation.

/// softwarezelle -* schichtenmodell -* code(101)

##### Abhängigkeiten definieren das Abstraktionsgefälle

Aus diesem Bild wird ersichtlich, wie Abstraktion und Abhängigkeit in Zusammenhang stehen: Abstraktion definiert einen Baum. Dessen Wurzel ist das Abstraktum, seine Blätter sind die Details, von denen damit abgesehen wird.

/// abstraktionsbaum mit abhängigkeiten; beispiel vllt mit "Gebäude" und dann "Wohnhaus", "Hochhaus", "Kirche" (102)

Das Abstrakte *besteht* aus dem Konkreten. Das Abstrakte ist somit vom Detail *abhängig*, das in ihm unter einem Gesichtspunkt zusammengefasst ist. Die Abhängigkeitsbeziehung weist vom Abstrakten zum Konkreten. In Richtung von Abhängigkeiten fällt mithin das Abstraktionsniveau.

##### Abstraktionen im Flow-Design

Flow-Design ist vor allem mit Abstraktionen nach zwei Gesichtspunkten beschäftigt:

* Aggregation: Zusammenfassen von *Ähnlichem*; das, was gemeinsam ist, definiert die Abstraktion.
* Komposition: Zusammenfassen von *Verschiedenem*; das Neue, das durch die gegenseitige Ergänzung des Verschiedenen entsteht, macht die Abstraktion aus.

/// abstraktionsbäume für aggregation (irgendwas gemeinsames wird rausgezogen) und komposition (irgendwas neues entsteht) (103)

Das Mittel zur Aggregation sind Module. Das Mittel zur Komposition sind Integration und Operation.

Aggregation erzeugt Überblick durch Ausblenden der Verschiedenartigkeit von Konkretem. Komposition erzeugt Neues durch Verschmelzen der Verschiedenartigkeit von Konkretem.

##### Fehlende Abstraktionen bei funktionaler Abhängigkeit

Vor diesem Hintergrund wird deutlich, dass Code, der nach dem Schichtenmodell strukturiert ist, Abhängigkeiten falsch im Sinne von Abstraktion einsetzt.

`Presentation.CountWords` als Wurzel des Abhängigkeitsbaums kann noch als Repräsentant des Ganzen auf höchster Abstraktionsebene angesehen werden; immerhin geht es um das Zählen von Worten in einer Datei.

Aber wie steht es mit `Business.CountWords`? Liegt die Methode auf einem niedrigeren Abstraktionsniveau? Tut sie dasselbe, wie die von ihr abhängige, nur eben konkreter? Ja, vielleicht könnte man das noch zugestehen. Es ja immerhin wieder um das Zählen von Worten in einer Datei; dieses Mal wird deren Zahl als Ergebnis geliefert, statt auf dem Bildschirm ausgegeben.

Und schließlich `Dataaccess.Load`: Ist damit die Problemlösung auf einem niedrigeren Abstraktionsniveau beschrieben? Nein. Hier geht es nicht mehr ums Zählen. Das ist deutlich sichtbar. Spätestens bei der Abhängigkeit der Geschäftslogik von der Datenzugriffslogik wird das Prinzip verletzt, dass Abhängigkeiten vom Abstrakten auf das Konkrete verweisen sollen.

Abstraktionsniveaus sind dadurch gekennzeichnet, dass auf einem bestimmten Niveau jeweils das Ganze zu sehen ist, nur eben in einem gewissen Detaillierungsgrad. Das ist offensichtlich bei `Dataaccess.Load` nicht der Fall und schon bei `Business.CountWords` wackelt das Bild.

Die Abhängigkeiten zwischen Methoden in einem Schichtenmodell beschreiben also keinen Abstraktionsbaum. Doch wie steht es mit dem Klassenbaum? Liegt eine `Presentation`-Klasse auf einem höheren Abstraktionsniveau als eine `Business`- oder `Dataaccess`-Klasse? Tut `Presentation` das in Aggregation, was `Business` oder `Dataccess` tun? Hier antwortet Flow-Design ebenfalls mit Nein.

Eine Geschäftslogik ist nicht dasselbe wie Datenzugriffslogik, nur eben auf einem anderen Abstraktionsniveau.

##### Dark Logic

Der Grund für diese "Missweisung" der Abhängigkeiten ist ihre funktionale Natur. Bäume funktionaler Abhängigkeiten enthalten *dark logic*, d.h. Logik, die Verhaltensrelevant ist, aber nicht im Modell repräsentiert. Das wird deutlich bei einer Darstellung des Lösungsansatzes für die Wortzählung als Datenflussdiagramm:

/// datenfluss für wortzählung (103)

Der Input von `Business.CountWords` passt noch zum Input von `Presentation.CoundWords`, aber der Output passt nicht. Was passiert mit der hinausfließenden Zahl der Worte? Genauso bei `Business.CoundWords` und `Dataaccess.Load`: Wie wird aus einer collection von Strings eine Zahl?

Beides ist nur zu erklären mit dark logic: In der abhängigen, integrierenden Funktionseinheit muss noch Logik stecken. Diese Logik widerspricht dem SLA:

```csharp
public static int CountWords(string filename) {
    var words = Dataaccess.Load(filename);
    var significantWords = words.Where(w => w.Length > 2);
    return significantWords.Count();
}
```

`Dataaccess.Load` liegt auf einem höheren Abstraktionsniveau als die folgenden Zeilen, weil die Methode selbst Logik zu einer neuen Funktionalität komponiert.

Laut IOSP kann das jedoch nicht sein.

Dark logic ist es, die Code schwer zu verstehen und schwer zu testen macht. Abhängigkeitsdiagrammen von Funktionen, die *funktional* abhängig sind, fehlt wesentliche Information, um zu verstehen, wie Verhalten hergestellt wird. Nicht, dass Logik explizit gezeigt werden müsste; damit würde der Abstraktionszweck gebrochen. Aber Logik muss explizit repräsentiert werden.

Das ist, wozu die Einhaltung des IOSP führt. Das IOSP macht dark logic sichtbar.

##### Stratified Design

Software strukturiert nach IOSP besteht nicht aus Schichten, sondern *Strata*. Jedes Stratum definiert dabei ein Abstraktionsniveau.

> "\[...\] expert engineers stratify complex designs. Each level is constructed as a stylized combination of interchangeable parts that are regarded as primitive at that level. The parts constructed at each level are used as primitives at the next level. Each level of a stratified design may be thought of as a specialized language with a variety of primitives and means of combination appropriate to that level of detail.", Abelson & Sussman in "Lisp: A Language for Stratified Design"

Das Schichtenmodell (oder auch eine Clean Architecture) sind keine Repräsentaten von *stratified design*. Lösungen existieren dort nicht auf unterschiedlichen Abstraktionsniveaus, weil Schichten/Schalten funktional von einander abhängig sind.

Anders jedoch im Flow-Design. Hier ein Datenfluss-Modell für das Wortzählungsproblem:

/// datenfluss für das wortzählungsproblem (104)

In diesem Modell existiert keine dark logic mehr. Die Lösung wird auf in zwei Strata komplett beschrieben:

/// zwei strata für das wortzählungsproblem: nur die wurzel / nur der flow (105)

Die gesamte Logik ist in Operationen verpackt.

Das höchste Abstraktionsniveau wurde durch das IOSP explizit eingeführt: die Klasse `CountWords` mit der Methode `Run`, die beide für Integration stehen. `CountWords` integriert die Aspekt-Klassen `Presentation`, `Business` und `Dataaccess`. `Run` integriert deren Methoden.

In [104]:
class CountWords {
    public static void Run(string filename) {
        var words = Dataaccess.Load(filename);
        var numberOfWords = Business_Stratified.CountWords(words);
        Presentation_Stratified.DisplayResult(filename, numberOfWords);
    }
}


class Presentation_Stratified {
    public static void DisplayResult(string filename, int numberOfWords) {
        Console.WriteLine($"Number of significant words in '{filename}': {numberOfWords}");
    }
}

class Business_Stratified {
    public static int CountWords(IEnumerable<string> words) {
        var significantWords = words.Where(w => w.Length > 2);
        return significantWords.Count();
    }
}


CountWords.Run("samples/poem.txt");

Number of significant words in 'samples/poem.txt': 72


Was vorher funktional abhängig war, steht jetzt in einer Sequenz im Datenfluss in `Run`. So ist es richtig, wenn es keine Unterschiede im Abstraktionsniveau gibt.

Im Beispiel macht die Geschäftslogik klar, wieviel plausibler das ist. Sie ist jetzt nicht mehr abhängig vom Datenzugriff. Warum sollte sie das auch sein? Welchen Grund gibt es, dass Domänenlogik sich Daten aus einer Ressource beschaffen sollte? Warum sollte Domänenlogik Kenntnis von einer Ressource haben?

Das Schichtenmodell erklärt das nicht. Die Clean Architecture erklärt das auch nicht, obwohl auch dort zur Laufzeit (!) noch innere Schalen äußere aufrufen.

Um den Kern einer Software wirklich von der Umwelt zu isolieren, darf es keine funktionalen Abhängigkeiten zwischen ihm und der Umwelt gehen. Funktionale Abhängigkeiten sind schlicht ein Widerspruch zum SLA. Mit ihnen würde die Geschäftslogik auf ein höheres Abstraktionsniveau gehoben als der Datenzugriff. Dort hat sie jedoch nichts zu suchen. Geschäftslogik ist nicht abstrakter als Datenzugriffslogik. Vielleicht ist sie zentraler, wichtiger, aber nicht abstrakter.

##### Strata verstanden als Sprachen

Die Lösung wird mit drei Sprache beschrieben:

* Die höchste Sprache besteht nur aus einem Wort: `Run`.
* Die niedrigere Sprache besteht aus drei Worten: `Load`, `CountWords` und `DisplayResult`.
* Die niedrigste Sprache besteht aus den Logik-Anweisungen in den Operationen.

Der Jo-Jo Entwurf von Flow-Design kann also als Sprachentwurf verstanden werden. Mit welchen sukzessive "primitiveren" - oder besser: allgemeineren - Sprachen lässt sich eine Lösung formulieren.

Die höchste Sprache bzw. die spezifischste besteht dabei immer nur aus einem Wort. Im Grunde ist es ein ["Sesam, öffne dich"](https://de.wikipedia.org/wiki/Ali_Baba) oder ["simsalabim"](https://de.wikipedia.org/wiki/Zauberspruch): das ganze Problem möge sich wie von Zauberhand durch Aufruf dieses einen Wortes in Luft auflösen.

Demgegenüber besteht die primitivste, allgemeinste Sprache immer aus allen Worten der zugrundgelegten Plattform aus Programmiersprache, Bibliotheken, Frameworks, APIs.

Und zwischen diesen beiden können beliebig viele weitere Sprachen eingezogen werden. Bei trivialen Problem mögen die höchste und primitivste ausreichen. Alle interessanten Probleme jedoch werden am besten durch Spezialsprachen gelöst, die aufeinander aufbauen.

Dazu ist es nicht nötig, *Domain Specific Languages (DSL)*, gar grafische, zu entwickeln. Simple Funktionen (und Module) genügen.

Im Flow-Design gleichen sich diese Sprachen alle in ihrer Syntax, die durch Datenflüsse (und nicht-funktionale Abhängigkeiten) vorgegeben ist. Das Vokabular jedoch ist zu (er)finden.

In Bezug auf ein etwas erweitertes Wortzählungsproblem ließen sich z.B. weitere Abstraktionsebenen mit eigenem Vokabular einziehen. Das könnte so aussehen:

/// neue datenflusshierarchie für wortzählung mit mehr integrationsebenen (106)

Im Code ist Logik nun in kleinere Einheiten verpackt. Die Funktionen sind feingranularer, das Vokabular also differenzierter (siehe Klassen der Geschäfts- und Datenzugriffslogik):

In [110]:
class CountWords_v2 {
    public static void Run(string filename) {
        var words = Dataaccess_Stratified.Load(filename);
        var numberOfWords = Business_Stratified_v2.CountWords(words);
        Presentation_Stratified.DisplayResult(filename, numberOfWords);
    }
}

class Business_Stratified_v2 {
    public static int CountWords(IEnumerable<string> words) {
        var significantWords = FilterShort(words);
        significantWords = FilterNumbers(significantWords);
        return significantWords.Count();
    }
    
    private static IEnumerable<string> FilterShort(IEnumerable<string> words)
        => words.Where(w => w.Length > 2);
    
    private static IEnumerable<string> FilterNumbers(IEnumerable<string> words) {
        return words.Where(IsNoNumber);
        
        bool IsNoNumber(string candidateWord)
            => int.TryParse(candidateWord, out var _) is false;
    }
}

class Dataaccess_Stratified {
    public static IEnumerable<string> Load(string filename) {
        var text = LoadText(filename);
        return SplitIntoWords(text);
    }
    
    private static string LoadText(string filename)
        => System.IO.File.ReadAllText(filename);
    
    private static IEnumerable<string> SplitIntoWords(string text)
        => text.Split(new[]{' ', '\t', '\n', '\r'}, StringSplitOptions.RemoveEmptyEntries);
}


CountWords_v2.Run("samples/poem.txt");

Number of significant words in 'samples/poem.txt': 72


Die Lösung besteht jetzt aus drei Strata:

1. `Run`
2. `Load`, `CountWords`, `DisplayResult`
3. `LoadText`, `SplitIntoWords`, `FilterShort`, `FilterNumbers`, `Count`, `DisplayResult`

Das Vokabular wechselt von Stratum zu Stratum. Die Funktionseinheiten werden allgemeiner und allgemeiner, d.h. ihr Einsatzbereich wird größer und größer. Je tiefer ein Stratum liegt, desto größer die potenzielle Wiederverwendbarkeit der Funktionseinheiten in darüberliegenden Strata.

Tendenziell werden Funktionseinheiten eines Stratums nur im direkt darüber liegenden benutzt. Gelegentlich greifen jedoch auch weiter oben liegende Strata darauf zu. Manches Vokabular ist dann so grundlegend, dass es sich durch mehrere Abstraktionsebenen durchzieht. Streng genommen könnte das als Verstoß gegen das SLA angesehen werden - doch im Einzelfall mag das besser verständlich sein als weitere Kapselungen in den einzelnen Strata.

Deshalb ist auch Logik hier und da in der Integration im Flow-Design geduldet. Logik ist das Vokabular des Terrains und sollte deshalb nicht in einem Stratum der Abstraktion auftauchen - außer eben in Einzelfällen. Beispiel dafür hier ist `Count` in `Business_Stratified_v2.CountWords`, deren Aufruf zum dritten Stratum gerechnet wird.

Unterhalb der Strata der Abstraktion liegt dann die Logik selbst. Ohne Operationen, Integrationen und Module ist sie quasi nackt. Jedes Detail ist sichtbar - aber die Lösung in dieser Weise entschleiert ist schwer verständlich und ihre Aspekte sind nicht getrennt testbar.

In [117]:
var filename = "samples/poem.txt";

var text = System.IO.File.ReadAllText(filename);
IEnumerable<string> words = text.Split(new[]{' ', '\t', '\n', '\r'}, StringSplitOptions.RemoveEmptyEntries);

words = words.Where(w => w.Length > 2);
words = words.Where(w => int.TryParse(w, out var _) is false);
var numberOfWords = words.Count();

Console.WriteLine($"Number of significant words in '{filename}': {numberOfWords}");

Number of significant words in 'samples/poem.txt': 72


##### Zusammenfassung

Code aufgebaut aus Strata wachsender Abstraktion um einen Kern aus Logik herum versprechen bessere Verständlichkeit und Veränderbarkeit. Das meinen nicht nur Abselson und Sussman, sondern auch Alan Kay, [der dabei ebenfalls von einer Hierarchie von Sprachen spricht](https://www.tele-task.de/lecture/video/2772/#t=3283).

/// schalen von strata mit logik im kern (107)

Nichts anderes geschieht ja auch ganz grundsätzlich, wenn man Lösungen mit Logik in einer modernen Programmiersprache formuliert. Selbst die reine Logik der obigen Wortzählungslogik lässt sich als abstraktes Stratum verstehen oberhalb des Stratums der .NET *Intermediate Language (IL)*, das wiederum auf dem Stratum des Prozessor-Maschinencodes liegt, der den tatsächlich ausgeführten Microcode abstrahiert.

/// schalen weiterführen unterhalb der logik mit microcode im kern (108)

Warum also diese erfolgreiche Methode nicht oberhalb der Logik ebenfalls anwenden?

Mit den Datenflüssen des Flow-Design geschieht das quasi natürlich durch die Anwendung von IOSP und PoMO. Dennoch ist die Vorstellung, in Strata von Sprache zu denken, ein gutes Hilfsmittel, um hierarchische Datenflüsse zu entwerfen.

Und zur Analyse von funktionalen Hierarchien erweist sich das Prinzip nützlich, dass Abhängigkeiten nur in Richtung des Abstraktionsgefälles zeigen sollen, also von hoher Abstraktion zu niedriger. MVC, Schichtenmodell, Clean Architecture oder auch alle Entwurfsmuster können daraufhin überprüft werden, ob ihre Abhängigkeiten dem entsprechen. Tun sie das nicht, ist aus Sicht von Flow-Design zumindest Vorsicht angezeigt. Allemal die Verständlichkeit mag dann geringer als möglich sein.

### 3-dimensionale Datenflüsse

Das erste Kennzeichen von Flow-Design Datenflüssen ist die Abwesenheit von Schleifen. Daten fließen nur downstream: "Vorwärts immer, rückwärts nimmer!"😉 ist das Motto.

Der Verzicht auf Schleifen ist eine "freiwillige Selbstbeschränkung". Er trägt dazu bei, dass Datenflüsse deklarativ sind, und erhöht die Verständlichkeit. Schleifen fordern mehr kognitiven Aufwand bei der Analyse von Code, weil sie dazu zwingen, etwas im Gedächtnis zu behalten, z.B. den Zustand einer Schleifenvariablen. Außerdem vergrößern sie das Risiko eines Bug z.B. durch 1-off Fehler:

```
var numbers = new[]{1,2,3};
var sum = 0;
for(var i=0; i<=numbers.Length; i++) //😱
    sum += numbers[i];
```

Moderne, abstraktere Sprachkonstrukte verringern zwar solches Risiko...

```
var numbers = new[]{1,2,3};
var sum = 0;
foreach(var n in numbers) //😇
    sum += numbers[i];
```

...und\/oder erhöhen die Verständlichkeit:

```
var numbers = new[]{1,2,3};
var sum = numbers.Sum(); //🥳
```

Letztlich bleiben Schleifen aber problematisch: Es sind Kontrollstrukturen, die zur Schachtelung einladen und unübersichtlich werden, wenn sie wachsen. Nicht in allen Fällen helfen moderne Abstraktionen. Deshalb gilt das Prinzip des Verzichts auf Schleifen in Datenflüssen.

Doch wie sieht es mit Fallunterscheidungen aus? Lassen sie sich auch vermeiden? Nein, Fallunterscheidungen definieren alternative Verhalten. Das ist notwendig und nicht zu vermeiden, auch wenn die Verhaltensproduktion mit Datenflüssen modelliert wird.

#### Kategorisierung

Eine häufige Anforderung im Rahmen der Verhaltensproduktion ist die Kategorisierung von Daten mit anschließend unterschiedlicher Behandlung je nach Kategorie. Beispiel: Eine Eingabe soll in eine Dezimalzahl konvertiert werden, wenn sie eine binäre Zahl ist, oder umgekehrt. Eine binäre Zahl liegt vor, wenn die Eingabe auf "b" endet.

Eine Lösung könnte mit Flow-Design so aussehen:

/// zahlenkonvertierung (109)

Neu ist hier, dass eine Funktionseinheit wie `Determine number system` mehrere Ausgänge haben kann. Die stehen für die Alternativen, dass der Input als binäre oder dezimale Zahl vorliegt. Jenachdem, zu welcher Kategorie der Input gehört, fließt die Verarbeitung entlang des einen oder des anderen Datenflussarms weiter.

In [129]:
class NumberConverter {
    public static string Convert(string number) {
        var result = "";
        DetermineNumberSystem(number,
            decimalNumber => result = ToBinary(decimalNumber),
            binaryNumber => result = ToDecimal(binaryNumber)
        );
        return result;
    }
    
    private static void DetermineNumberSystem(string number, Action<string> onIsDecimal, Action<string> onIsBinary) {
        if (number.EndsWith("b"))
            onIsBinary(number.Substring(0, number.Length-1));
        else
            onIsDecimal(number);
    }
    
    
    private static string ToBinary(string number)
        => System.Convert.ToString(int.Parse(number), 2);
    
    private static string ToDecimal(string number)
        => System.Convert.ToInt32(number,2).ToString();
}


display($"10b={NumberConverter.Convert("10b")}");
display($"10={NumberConverter.Convert("10")}");

10b=2

10=1010

Mehrere parallele Datenflüsse fügen den bisherigen 2-dimensionalen Modellen eine dritte Dimension hinzu:

/// 3 dimensionen in einem hierarchischen datenfluss kennzeichnen (komposition, produktion.linear und produktion.parallel) (109a)

Wie schon vorher bei den 1-dimensionalen Datenflüssen gilt auch hier: alle Funktionseinheiten arbeiten grundsätzlich unabhängig voneinander. Nacheinander geschaltete Funktionseinheiten können gleichzeitig aktiv sein, parallel geschaltete ebenfalls. Dass das oft nicht der Fall ist, weil nicht nötig wie im obigen Beispiel, tut dem keinen Abbruch. 3-dimensionale Datenflüsse erlauben die modellierung synchroner wie asynchroner, gar paralleler Verarbeitung.

##### Übersetzungsalternativen

Auffallend an den alternativen Ausgängen der Funktionseinheit `DetermineNumberSystem` ist, dass Output auf beiden als stream herausfließt. Das ist immer der Fall, wenn Output-Nachrichten entstehen können oder auch nicht. Wenn der Input eine dezimale Zahl darstellt, dann fließt Output auf dem zugehörigen Port heraus - und beim anderen nicht. Und umgekehrt. Nur mit streams ist es möglich, keinen Output zu erzeugen.

Das führt dann dazu, dass in der Übersetzung continuations zum Einsatz kommen. Nur mit ihnen lässt sich der Fall ohne Output für eine Alternative sauber implementieren.

Alternativ wäre in einer Sprache wie C# mit Tupel-Unterstützung zwar auch folgende Übersetzung möglich:

In [131]:
class NumberConverter_v2 {
    public static string Convert(string number) {
        var categorization = DetermineNumberSystem(number);
        if (categorization.isBinary)
            return ToDecimal(categorization.normalizedNumber);
        else
            return ToBinary(categorization.normalizedNumber);
    }
    
    private static (bool isBinary, string normalizedNumber) DetermineNumberSystem(string number) {
        if (number.EndsWith("b"))
            return (true, number.Substring(0, number.Length-1));
        else
            return (false, number);
    }
    
    
    private static string ToBinary(string number)
        => System.Convert.ToString(int.Parse(number), 2);
    
    private static string ToDecimal(string number)
        => System.Convert.ToInt32(number,2).ToString();
}


display($"10b={NumberConverter_v2.Convert("10b")}");
display($"10={NumberConverter_v2.Convert("10")}");

10b=2

10=1010

Doch damit würde die Integration `Convert` durch eine Kontrollstruktur verschmutzt. Das ist zwar nur minimal... dennoch sollte man sich dessen bewusst sein.

Im allgemeinen Fall würde diese Variante die Kategorie über einen `enum`-Datentyp zurückliefern.

Oder sie könnte sogar noch weitergehen und die downstream Verarbeitung selbst dynamisch bestimmen. Hierin könnte man zwar einen Widerspruch zum PoMO sehen... doch wenn Kategorisierung und Mapping auf eine Verarbeitung getrennt sind, mag das der Verständlichkeit und Testbarkeit nicht abträglich, sondern sogar zuträglich sein.

/// kategorisierung mit converter bestimmung (110)

Jetzt gibt es keine Continuations mehr im Code. Jetzt gibt es keine unsauberen Kontrollstrukturen mehr in Integrationen.

In [136]:
class NumberConverter_v3 {
    public static string Convert(string number) {
        var conversion = ChooseConverter(number);
        return conversion.convert(conversion.normalizedNumber);
    }
    
    private static (Func<string,string> convert, string normalizedNumber) ChooseConverter(string number) {
        var categorization = DetermineNumberSystem(number);
        var convert = PickConverter(categorization.isBinary);
        return (convert, categorization.normalizedNumber);
    }
    
    private static (bool isBinary, string normalizedNumber) DetermineNumberSystem(string number) {
        if (number.EndsWith("b"))
            return (true, number.Substring(0, number.Length-1));
        else
            return (false, number);
    }
    
    private static Func<string,string> PickConverter(bool isBinary) => isBinary switch {
        true => ToDecimal,
        false => ToBinary
    };
    
    
    private static string ToBinary(string number)
        => System.Convert.ToString(int.Parse(number), 2);
    
    private static string ToDecimal(string number)
        => System.Convert.ToInt32(number,2).ToString();
}


display($"10b={NumberConverter_v3.Convert("10b")}");
display($"10={NumberConverter_v3.Convert("10")}");

10b=2

10=1010

Dem Anfänger in Sachen Flow-Design fällt es gewöhnlich nicht leicht, sich vorzustellen, dass Kontrollstrukturen und insbesondere Fallunterscheidungen wirklich nur in den Blättern des Zerlegungsbaumes stehen. Das ist verständlich, weil es so ungewohnt ist. In den Bäumen tiefer funktionaler Abhängigkeiten ist das undenkbar und scheint auch gar nicht nötig.

Doch gerade hierin liegt das Geheimnis der Verständlichkeit und Testbarkeit von Code, der mit Flow-Design modelliert und IOSP/PoMO folgend übersetzt wird! Deshalb sollten alle Anstrengungen unternommen werden, Schleifen und Fallunterscheidungen wirklich in die Operation-Funktionen hinunter zu drücken. Es ist in jedem Fall machbar und ist der Default im Flow-Design. Abweichungen davon sollten sehr bewusst eingegangen werden; sie lassen dark logic entstehen und verwässern das stratified design.

#### Fehler

Ein weiterer typischer Fall für parallele Datenflüsse stellen Fehler dar. Da gibt es einerseits den Datenfluss für den "happy day", d.h. wenn alles wie erwartet läuft. Und andererseits gibt es den Datenfluss für den "rainy day", d.h. den Fehlerfall.

Ein typisches Szenario ist die Validation: Wenn ankommende Daten bestimmten Kriterien genügen, scheint die Sonne, ansonsten regnet es. Gutfall und Fehlerfall zu unterscheiden ist ein nicht zu vernachlässigender Teil jedes Modells. Im Flow-Design geschieht das wieder mit mehreren Ausgängen und parallelen Datenflüssen.

Als Beispiel die Addition mehrerer durch Leerzeichen getrennter Zahlen. Sind alle Zahlen valide, kann die Summe gebildet und ausgegeben werden. Falls nicht, gibt es eine Fehlermeldung.

/// modell für die zahlenaddition mit fehlerfall (111)

Gutfall oder Fehlerfall sind also im Grunde nur zwei Kategorien. Deren Bestimmung erfolgt oft jedoch nicht in einer speziellen Funktionseinheit zur Kategorisierung wie bei der Zahlenwandlung oben, sondern ist ein Beiprodukt. Hier nimmt z.B. `Parse` den Gutfall an, ist jedoch darauf vorbereitet, dass es zum Fehlerfall kommt.

In [142]:
public class StringMath {
    public static void Sum(string source, Action<int> onOk, Action<string> onError) {
        Parse(source,
             numbers => {
                 var sum = numbers.Sum();
                 onOk(sum);
             },
             erroneousNumber => {
                 var error = $"Cannot sum numbers! '{erroneousNumber}' is not a number.";
                 onError(error);
             }
         );
    }
    
    private static void Parse(string source, Action<int[]> onSuccess, Action<string> onFailure) {
        var candidateNumbers = source.Split(' ');
        var numbers = new List<int>();
        foreach(var cn in candidateNumbers)
            if (int.TryParse(cn, out var n))
                numbers.Add(n);
            else {
                onFailure(cn);
                return;
            }
        onSuccess(numbers.ToArray());
    }
}


var input = "1 2 3";
StringMath.Sum(input,
    sum => display($"sum of {input} = {sum}"),
    error => display($"Error: {error}")
);

input = "1 foo 3";
StringMath.Sum(input,
    sum => display($"sum of {input} = {sum}"),
    error => display($"Error: {error}")
);

sum of 1 2 3 = 6

Error: Cannot sum numbers! 'foo' is not a number.

Das der Gutfall der wünschenswerte und meist interessantere und hoffentlich auch der häufigere ist, könnte man versucht sein, die Asymmetrie zwischen Gutfall und Fehlerfall im Code auszudrücken. Warum nicht den Output des Gutfalls über das Funktionsresultat zurückliefern und den Fehlerfall in einer continuation abhandeln?

```csharp
private static int[] Parse(string source, Action<string> onFailure) {
    var candidateNumbers = source.Split(' ');
    var numbers = new List<int>();
    foreach(var cn in candidateNumbers)
        if (int.TryParse(cn, out var n))
            numbers.Add(n);
        else {
            onFailure(cn);
            return null; //😱
        }
    return numbers.ToArray();
}
```

Das Problem ist der Rückgabewert der Funktion im Fehlerfall. Warum sollte der wie hier `null` sein? Darauf muss die Gutfall-Bearbeitung vorbereitet sein. Das vergrößert die Komplexität des Codes.

Einer symmetrische Behandlung ist der Vorzug zu geben. Wenn die nicht durch continuations stattfinden soll, dann wäre es besser, die Alternative Klassifizierungsübersetzung zum Einsatz zu bringen:

```csharp
private static (bool success, int[] numbers, string error) Parse(string source) {
    var candidateNumbers = source.Split(' ');
    var numbers = new List<int>();
    foreach(var cn in candidateNumbers)
        if (int.TryParse(cn, out var n))
            numbers.Add(n);
        else 
            return (false, new int[0], cn);
    return (true, numbers.ToArray(), "");
}
```

Das ist ein anerkanntes Muster neben der einfacheren Variante einer `Try`-Funktion, z.B. `bool TryParse(string source, out int[] numbers)`.

Ohne continuations rutscht aber in jedem Fall Kontrollstruktur-Logik in die Integration, in der eine solche Funktion aufgerufen wird. Es ist also sensibel abzuwägen.

#### Ausnahmen

Es gibt erwartete Fehler und unerwartet. Letztere werden im Notfall durch Exceptions angezeigt. Die können wie rainy-day Fehler modelliert werden und insofern erwartet aussehen. Oder man benutzt eine spezielle Datenflussnotation, um anzuzeigen, dass Ausnahme nicht näher bestimmt aus einer Reihe von Funktionseinheiten herausspringen können.

Als Beispiel ein Datenfluss der eine Reihe von Dateien verarbeitet: Sie werden gelesen, analysiert und das Analyseergebnis am Ende ausgegeben. In jedem der Schritte könnte eine Ausnahme geworfen werden. Das wird nicht erwartet, aber die Möglichkeit besteht. Deshalb gibt es keine speziellen Fehlerausgänge bei allen Funktionseinheiten, sondern sie werden in einen "Kontext" gesteckt, der allgemein darauf reagiert. (Die Funktionale Programmierung könnte dazu wohl auch *Monade* sagen.)

/// datenfluss mit lesen, analysieren, summieren, anzeigen (112)

Für den möglichen, aber nicht wahrscheinlichen Fall einer Ausnahme jede Funktionseinheit mit einem speziellen Output-Port dafür zu versehen, würde die Komplexität des Modells und des Codes stark erhöhen.

In [152]:
class FileStats {
    public static void CountLines(string path) {
        try {
            (int numberOfFiles, int totalNumberOfLines) accumulator = (0,0);
            EnumerateFiles(path,
                filename => {
                    var lines = System.IO.File.ReadAllLines(filename);
                    accumulator.numberOfFiles += 1;
                    accumulator.totalNumberOfLines += lines.Length;
                });
            DisplayResult(accumulator);
        }
        catch(Exception ex) {
            DisplayException(ex);
        }
    }
    
    
    private static void EnumerateFiles(string path, Action<string> onFile) {
        foreach(var filename in System.IO.Directory.GetFiles(path, "*.*", System.IO.SearchOption.AllDirectories))
            onFile(filename);
    }
    
    
    private static void DisplayResult((int numberOfFiles, int totalNumberOfLines) result)
        => Console.WriteLine($"{result.numberOfFiles} files with {result.totalNumberOfLines} lines in total");
    
    private static void DisplayException(Exception ex)
        => Console.WriteLine($"Exception: {ex.Message}");
}


FileStats.CountLines("../entwurf");
FileStats.CountLines("../foo");

23 files with 28384 lines in total
Exception: Could not find a part of the path '/Users/ralfw/Projects/jupyter-notebooks.github/Notebooks/flow-design-tutorial/foo'.


Die Übersetzung des Kontextes mit `try`-`catch` ist naheliegend. Dass diese Kontrollstruktur jedoch in der Integration liegt, ist ein Widerspruch zum IOSP. Da Verständlichkeit und Testbarkeit dadurch jedoch nicht beeinträchtigt werden, ist das ausnahmsweise jedoch kein Problem.

Wer jedoch auf Nummer sicher gehen will, insbesondere wenn noch weitere Logik in beiden Fällen hinzukommen soll, der lagert die Kontrollstruktur aus in eine eigene Methode wie die folgende:

In [154]:
class Exceptionhandling {
    public static void Try(Action tryThis, Action<Exception> onException) {
        try {
            display($"trying...");
            tryThis();
            display($"succeeded!");
        }
        catch(Exception ex) {
            display($"failed!");
            onException(ex);
        }
    }
}

Zusätzliche Logik ist hier angedeutet durch die Protokollnachrichten. Sie würden die eigentliche Logik verrauschen und sind ein einem "Kontext" als separater Aspekt gut ausgelagert.

Die Lösung sieht dann nur wenig anders aus, enthält aber keine Logik mehr in der Integration `CountLines`:

In [155]:
class FileStats_v2 {
    public static void CountLines(string path) {
        Exceptionhandling.Try(
            () => Analyze(path),
            DisplayException
        );
    }
    
    private static void Analyze(string path) {
        (int numberOfFiles, int totalNumberOfLines) accumulator = (0,0);
        EnumerateFiles(path,
            filename => {
                var lines = System.IO.File.ReadAllLines(filename);
                accumulator.numberOfFiles += 1;
                accumulator.totalNumberOfLines += lines.Length;
            });
        DisplayResult(accumulator);
    }
    
    private static void EnumerateFiles(string path, Action<string> onFile) {
        foreach(var filename in System.IO.Directory.GetFiles(path, "*.*", System.IO.SearchOption.AllDirectories))
            onFile(filename);
    }
    
    
    private static void DisplayResult((int numberOfFiles, int totalNumberOfLines) result)
        => Console.WriteLine($"{result.numberOfFiles} files with {result.totalNumberOfLines} lines in total");
    
    private static void DisplayException(Exception ex)
        => Console.WriteLine($"Exception: {ex.Message}");
}


FileStats_v2.CountLines("../entwurf");
FileStats_v2.CountLines("../foo");

trying...

23 files with 28936 lines in total


succeeded!

trying...

failed!

Exception: Could not find a part of the path '/Users/ralfw/Projects/jupyter-notebooks.github/Notebooks/flow-design-tutorial/foo'.


#### Flüsse teilen und zusammenführen

Mehrere Datenflüsse können nicht nur alternativ durchflossen werden, sondern auch gleichzeitig. Das ist z.B. der Fall, wenn Input Teile enthält, die unabhägig von einander verarbeitet werden können. Von der Datenflussnotation her könnte solche Verarbeitung zwar mit einem 1-dimensionalen Datenfluss modelliert werden, aber klarer ist es, wenn Parallelität auch visuell ausgedrückt wird.

##### Split/Join

Als Beispiel mag die Analyse von Dateien in Bezug auf zwei Aspekte dienen: einerseits soll ihre Zeilenzahl bestimmt werden, andererseits ihre Wortanzahl. Beide Informationen werden anschließend zu einem Ergebnis zusammengeschnürt.

/// datei laden - split für zeilenzählung / wortzählung, join für ergebniszusammenführung (mit stats) datenstruktur (113)

Nach dem Laden der Datei wird der Datenfluss geteilt (split), beide Verarbeitungszweige laufen unabhängig, am Ende werden deren Teilergebnisse zu einem Gesamtergebnis zusammengeführt (join).

Sofern eine Zusammenführung unaufwändig ist, muss keine eigens benannte Funktionseinheit dafür modelliert werden. Es reicht das join-Symbol wie hier gezeigt.

In [164]:
class FileAnalysis {
    public class Result {
        internal Result(int numberOfLines, int numberOfWords) {
            NumberOfLines = numberOfLines;
            NumberOfWords = numberOfWords;
        }
        
        public int NumberOfLines {get;}
        public int NumberOfWords {get;}
    }
    
    
    public static Result Analyze(string filename) {
        var text = Load(filename);
        return new Result(
            CountLines(text),
            CountWords(text)
        );
    }
    
    
    private static string Load(string filename)
        => System.IO.File.ReadAllText(filename);
    
    
    private static int CountLines(string text) {
        var lines = text.Split('\n');
        return lines.Length;
    }
    
    private static int CountWords(string text) {
        var words = text.Split(new[]{' ', '\t', '\n', '\r'}, StringSplitOptions.RemoveEmptyEntries);
        return words.Count();
    }
}


display(FileAnalysis.Analyze("samples/poem.txt"));

NumberOfLines NumberOfWords 15 79

Was im Modell noch explizit ist, geschieht im Code dann fast unmerklich:

* Der split findet keine spezielle Übersetzung; dieselben Daten werden einfach als aktuelle Parameter zweier Funktionsaufrufe benutzt.
* Die Parallelität der Datenflüsse ist nicht deutlich zu sehen, weil der lineare Quelltext das nicht erlaubt. Nur aus der Benutzung derselben Daten als Input für aufeinander folgende Funktionsaufrufe ist Parallelität herauszulesen.
* Der join besteht schlicht in der Nutzung beider Teilergebnisse als Parameter eines weiteren Funktionsaufrufs.

In anderen Fällen mögen split und join hingegen sowohl im Modell wie im Code deutlicher zu sehen sein.

/// beispiel für split und join funktionseinheiten (fluss dazwischen nur andeuten) (114)

Zweierlei ist dabei dann zu bedenken:

* Split: Fließt Output aus den Ports optional/alternativ oder stets? Optionalität fordert die Modellierung als stream und legt eine Implementation als continuation nahe.
* Join: Soll Verarbeitung bei jedem Eintreffen von Input stattfinden, egal auf welchem Port er hineinfließt, oder nur, wenn an allen Ports Input anliegt? Letzteres lässt sich leicht in mehrere Funktionsparameter übersetzen wie oben beim `Result`-Konstruktor zu sehen. Ersteres legt eine Übersetzung in mehrere separate Funktionen einer Klasse nahe.

##### Scatter/Gather

Ein Beispiel für explizite splits und joins sind echte Parallelverarbeitungsszenarien, d.h. Übersetzungen, in denen mehrere Threads zum Einsatz kommen.

Beispiel: Dateien sollen parallel analysiert werden. Anschließend werden alle Analyseergebnisse zu einem zusammengeführt. Die Analyse kann so simpel sein wie die Zählung der Zeilen je Datei.

/// paralleler datenfluss für dateizeilenzählung mit join am ende (115)

Zunächst findet hier eine unbekannt große Aufsplittung des Datenflusses in parallele/konkurrente statt (scatter). Das kann mit einem stream geschehen, bei dem für jedes Element ein Thread gestartet wird.

Wie lange die einzelnen Analysen dauern und wieviele es sind, ist bei der Akkumulation des Ergebnisses jedoch nicht bekannt (gather). Wann soll also deren Output ausfließen? Wie weiß die Akkumulation, dass das letzte Analyseergebnis eingetroffen ist? Das kann implizit mittels einer Framework-Technologie geschehen (z.B. ein `Task.WaitAll` im .NET Framework) oder explizit durch Vergleich der Zahl der eingetroffenen Ergebnisse mit der zu erwartenden Zahl, die von scatter gemeldet wird. Das Modell zeigt den zweiten Ansatz für gather, um technologieneutraler und expliziter zu sein.

In [12]:
using System.Threading;

class ConcurrentFileAnalysis {
    public static (int numberOfFiles, int totalNumberOfLines) Analyze(string path) {
        var acc = new Accumulator();
        
        var filenames = CompileFiles(path);
        Scatter(filenames,
            filename => {
                var numberOfLines = AnalyzeFile(filename);
                acc.Add(numberOfLines);
            },
            acc.SetNumberOfValuesToExpect
        );
            
        return acc.Gather();
    }
    
    private static string[] CompileFiles(string path) {
        var filenames = System.IO.Directory.GetFiles(path, "*.ipynb", System.IO.SearchOption.AllDirectories);
        return filenames.Where(filename => filename.IndexOf("/.") < 0).ToArray();
    }
    
    private static void Scatter(string[] filenames, Action<string> onFilename, Action<int> onAllScattered) {
        foreach(var filename in filenames) {
            var th = new Thread(() => onFilename(filename));
            th.Start();
            //onFilename(filename);
        }
        onAllScattered(filenames.Length);
    }
    
    private static int AnalyzeFile(string filename) {
        display($"  analyzing {filename}");
        var lines = System.IO.File.ReadAllLines(filename);
        return lines.Length;
    }
}

private class Accumulator {
    private object _lock = new Object();
    private AutoResetEvent _are = new AutoResetEvent(false);

    private int _numberOfValuesToExpect = 0;
    private int _numberOfValuesReceived = 0;

    private int _total;

    
    public void SetNumberOfValuesToExpect(int numberOfValuesToExpect) {
        lock(_lock) {
            _numberOfValuesToExpect = numberOfValuesToExpect;
            
            if (_numberOfValuesReceived >= _numberOfValuesToExpect)
                _are.Set();
        }
    }

    public void Add(int value) {
        lock(_lock) {
            _numberOfValuesReceived++;
             _total += value;
            
            if (_numberOfValuesToExpect > 0 && _numberOfValuesReceived >= _numberOfValuesToExpect)
                 _are.Set();
        }
    }
    
    public (int numberOfValues, int totalValue) Gather() {
        _are.WaitOne();
        return (_numberOfValuesToExpect, _total);
    }
}


var result = ConcurrentFileAnalysis.Analyze("..");
display($"{result.numberOfFiles} files with {result.totalNumberOfLines} lines total");

  analyzing ../prozess.ipynb

  analyzing ../anforderungen.ipynb

  analyzing ../anforderung-logik-lücke.ipynb

  analyzing ../index.ipynb

  analyzing ../entwurf/entwurf_1.ipynb

  analyzing ../codierung/arbeitsorganisation.ipynb

  analyzing ../analyse/slicing.ipynb

  analyzing ../anatomie/radikale_oo.ipynb

  analyzing ../anatomie/ioda.ipynb

  analyzing ../anatomie/dimensionen.ipynb

10 files with 5601 lines total



datenströme explizit beenden (hier ein thema, weil bei 3D datenflüssen eher continuations zum einsatz kommen)
oder auch parallelverarbeitung: fork join/wait, scatter/gather (dateien parallel verarbeiten).
ende implizit (IEnum) oder EOS element oder signal (dass ende erreicht oder wann ende erreicht (anzahl zu erwartender inputs))